File: 05-telnyx-media-streaming/python/outbound_call.py
import os

import httpx
from dotenv import load_dotenv

load_dotenv(".env.local")


def make_call():
    # Places an outbound call with the Telnyx Call Control API (Voice API
    # application). Streaming is NOT started here: bidirectional return audio
    # must be set up with streaming_start AFTER the call is answered, so we
    # point Telnyx's call webhooks at /call-control and server.py starts the
    # stream on the call.answered event.
    response = httpx.post(
        "https://api.telnyx.com/v2/calls",
        headers={"Authorization": f"Bearer {os.environ['TELNYX_API_KEY']}"},
        json={
            "connection_id": os.environ["TELNYX_CONNECTION_ID"],
            "to": os.environ["CUSTOMER_PHONE_NUMBER"],
            "from": os.environ["TELNYX_PHONE_NUMBER"],
            # Per-call webhook override, so you don't have to configure the
            # application's webhook URL in the portal.
            "webhook_url": f"{os.environ['NGROK_URL']}/call-control",
        },
    )

    if response.status_code >= 400:
        print(f"Error making call: {response.status_code} {response.text}")
        return

    call_control_id = response.json()["data"]["call_control_id"]
    print(f"Call initiated with control ID: {call_control_id}")


if __name__ == "__main__":
    make_call()

File: 05-telnyx-media-streaming/node/outbound-call.ts
import {
  customerPhoneNumber,
  ngrokUrl,
  requireOutboundEnvVars,
  telnyxApiKey,
  telnyxConnectionId,
  telnyxPhoneNumber,
} from "./env-vars";

requireOutboundEnvVars();

// Places an outbound call with the Telnyx Call Control API (Voice API
// application). Streaming is NOT started here: bidirectional return audio has
// to be set up with streaming_start AFTER the call is answered, so we point
// Telnyx's call webhooks at /call-control and server.ts starts the stream on
// the call.answered event.
async function makeCall() {
  const response = await fetch("https://api.telnyx.com/v2/calls", {
    method: "POST",
    signal: AbortSignal.timeout(5000),
    headers: {
      Authorization: `Bearer ${telnyxApiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      connection_id: telnyxConnectionId,
      to: customerPhoneNumber,
      from: telnyxPhoneNumber,
      // Per-call webhook override, so you don't have to configure the
      // application's webhook URL in the portal.
      webhook_url: `${ngrokUrl}/call-control`,
    }),
  });

  if (!response.ok) {
    console.error(
      `Error making call: ${response.status} ${response.statusText}`,
      await response.text(),
    );
    return;
  }

  const { data } = (await response.json()) as {
    data: { call_control_id: string };
  };
  console.log(`Call initiated with control ID: ${data.call_control_id}`);
}

makeCall();
